godel_process_execution/src/keyence_process_service.cpp

```diff
@@ -0,0 +1,148 @@
+
+  // Create this process execution server
+  server_ = nh.advertiseService<KeyenceProcessExecutionService,
+                                godel_msgs::KeyenceProcessExecution::Request,
+                                godel_msgs::KeyenceProcessExecution::Response>
+            (name, &godel_process_execution::KeyenceProcessExecutionService::executionCallback, this);
+}
+
+bool godel_process_execution::KeyenceProcessExecutionService::executionCallback(godel_msgs::KeyenceProcessExecution::Request& req,
+                                                                                godel_msgs::KeyenceProcessExecution::Response& res)
+{
+  using moveit_msgs::ExecuteKnownTrajectory;
+  using simulator_service::SimulateTrajectory;
+
+  if (req.simulate)
+  {
+    return simulateProcess(req);
+  }
+  else
+  {
+    if (req.wait_for_execution)
+    {
+      return executeProcess(req);
+    }
+    else
+    {
+      boost::thread(&godel_process_execution::KeyenceProcessExecutionService::executeProcess, this, req);
+      return true;
+    }
+  }
+}
+
+bool godel_process_execution::KeyenceProcessExecutionService::executeProcess(godel_msgs::KeyenceProcessExecution::Request& req)
+{
+  // Check for keyence existence
+  if (!keyence_client_.exists())
+  {
+    ROS_WARN_STREAM("Keyence ROS server is not available on service " << keyence_client_.getService());
+    return false;
+  }
+
+  godel_msgs::TrajectoryExecution srv_approach;
+  srv_approach.request.wait_for_execution = true;
+  srv_approach.request.trajectory = req.trajectory_approach;
+
+  godel_msgs::TrajectoryExecution srv_process;
+  srv_process.request.wait_for_execution = true;
+  srv_process.request.trajectory = req.trajectory_process;
+
+  godel_msgs::TrajectoryExecution srv_depart;
+  srv_depart.request.wait_for_execution = true;
+  srv_depart.request.trajectory = req.trajectory_depart;
+
+  if (!real_client_.call(srv_approach))
+  {
+    ROS_WARN("Execution client unavailable or unable to execute approach trajectory.");
+    return false;
+  }
+
+  keyence_driver::ChangeProgram keyence_srv;
+  keyence_srv.request.program_no = KEYENCE_PROGRAM_LASER_ON;
+
+  if (!keyence_client_.call(keyence_srv))
+  {
+    ROS_WARN_STREAM("Unable to activate keyence (program " << KEYENCE_PROGRAM_LASER_ON << ").");
+     return false;
+  }
+
+  if (!real_client_.call(srv_process))
+  {
+    ROS_WARN("Execution client unavailable or unable to execute process trajectory.");
+    return false;
+  }
+
+  // Turn keyence off
+  keyence_srv.request.program_no = KEYENCE_PROGRAM_LASER_OFF;
+  if (!keyence_client_.call(keyence_srv))
+  {
+    ROS_WARN_STREAM("Unable to de-activate keyence (program " << KEYENCE_PROGRAM_LASER_OFF << ").");
+    return false;
+  }
+
+  if (!real_client_.call(srv_depart))
+  {
+    ROS_WARN("Execution client unavailable or unable to execute departure trajectory.");
+    return false;
+  }
+
+  return true;
+}
```

Comment:
Lots of warning messages that feel like candidates for converting into error messages.  If none of these conditions prevent the process from completing then they should probably stay as warnings.